File: frontend/src/services/api/agent.ts
import { getDefaultStore } from 'jotai'

import { apiV1Delete, apiV1Get, apiV1Post, apiV1Put } from '@/services/client'

import { ACCESS_TOKEN_KEY } from '@/utils/store'
import { configAPIBaseAtom } from '@/utils/store/config'

import type { IResponse } from '../types'

// ──────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────

export type AgentSurface = 'user' | 'admin'

export interface AgentSSEEvent {
  event:
    | 'agent_run_started'
    | 'agent_status'
    | 'agent_handoff'
    | 'thinking'
    | 'tool_call'
    | 'tool_result'
    | 'message'
    | 'confirmation_required'
    | 'tool_call_started'
    | 'tool_call_completed'
    | 'tool_call_confirmation_required'
    | 'parameter_review'
    | 'final_answer'
    | 'error'
    | 'done'
  data: unknown
}

export type AgentOrchestrationMode = 'single_agent' | 'ask'

export interface AgentSession {
  sessionId: string
  title: string
  source?: 'chat' | 'ops_audit' | 'system'
  messageCount: number
  lastOrchestrationMode?: AgentOrchestrationMode
  pinnedAt?: string | null
  createdAt: string
  updatedAt: string
}

export interface AgentTurn {
  id: number
  turnId: string
  sessionId: string
  requestId?: string
  orchestrationMode: AgentOrchestrationMode
  rootAgentId?: string
  status: string
  finalMessageId?: number | null
  metadata?: unknown
  startedAt: string
  endedAt?: string | null
  createdAt: string
  updatedAt: string
}

export interface AgentEvent {
  id: number
  turnId: string
  sessionId: string
  agentId: string
  parentAgentId?: string
  agentRole: string
  eventType: string
  eventStatus?: string
  title?: string
  content?: string
  metadata?: unknown
  sequence: number
  startedAt?: string
  endedAt?: string
  createdAt: string
}

export interface AgentConfigSummary {
  defaultOrchestrationMode: 'single_agent'
}

export interface AgentMessage {
  id: string
  sessionId?: string
  role: 'user' | 'assistant'
  content: string
  metadata?: unknown
  createdAt: string
}

export interface AgentToolCall {
  id: string
  turnId?: string
  toolCallId?: string
  agentId?: string
  agentRole?: string
  source?: 'backend' | 'local'
  toolName: string
  toolArgs?: unknown
  toolResult?: unknown
  resultStatus: string
  userConfirmed?: boolean | null
  createdAt: string
}

export interface AgentConfirmationFieldOption {
  value: string
  label: string
}

export interface AgentConfirmationField {
  key: string
  label: string
  type: 'text' | 'textarea' | 'number' | 'select'
  required?: boolean
  description?: string
  placeholder?: string
  defaultValue?: unknown
  options?: AgentConfirmationFieldOption[]
}

export interface AgentConfirmationForm {
  title?: string
  description?: string
  submitLabel?: string
  fields?: AgentConfirmationField[]
}

export interface AgentConfirmResponseData {
  status: string
  result?: unknown
  message?: string
}

type AgentPageContext = {
  route?: string
  url: string
  jobName?: string
  jobStatus?: string
  nodeName?: string
  entryPoint?: 'default' | 'node_analysis'
  surface?: AgentSurface
}

export interface AgentChatRequest {
  sessionId: string | null
  requestId?: string | null
  message: string
  orchestrationMode?: 'single_agent'
  pageContext: AgentPageContext
  clientContext?: {
    locale?: string
    timezone?: string
  }
}

export interface AgentAskRequest {
  sessionId: string | null
  requestId?: string | null
  message: string
  jobName?: string
  pageContext: AgentChatRequest['pageContext']
  clientContext?: AgentChatRequest['clientContext']
}

export interface AgentResumeRequest {
  confirmId: string
}

export interface ParameterReviewPayload {
  reviewId: string
  scenario: string
  complexity: 'simple' | 'complex'
  step: number
  totalSteps: number
  title: string
  description: string
  parameters: Array<{
    key: string
    label: string
    value: unknown
    source: 'recommended' | 'default' | 'user'
    editable: boolean
    type: 'text' | 'number' | 'select' | 'textarea'
    options?: Array<{ label: string; value: string }>
    constraints?: { min?: number; max?: number }
    hint?: string
  }>
}

// ──────────────────────────────────────────────
// SSE streaming connection
// ──────────────────────────────────────────────

function getAgentStreamBaseURL() {
  const store = getDefaultStore()
  const apiBase = store.get(configAPIBaseAtom)
  return `${apiBase ?? ''}/api`
}

function getAgentStreamError(parsed: unknown, fallback: string) {
  if (typeof parsed === 'string') return parsed
  if (typeof parsed === 'object' && parsed !== null) {
    const body = parsed as { message?: string; msg?: string }
    return body.message || body.msg || fallback
  }
  return fallback
}

function connectAgentStream<TBody>(
  path: string,
  body: TBody,
  options: {
    noBodyMessage: string
    fallbackStreamError: string
    onEvent: (event: AgentSSEEvent) => void
    onError: (error: Error) => void
    onDone: () => void
    onSessionId?: (sessionId: string) => void
  }
): AbortController {
  const controller = new AbortController()
  const token = localStorage.getItem(ACCESS_TOKEN_KEY)
  let doneDispatched = false

  const dispatchDone = () => {
    if (doneDispatched) return
    doneDispatched = true
    options.onDone()
  }

  const dispatchStreamError = (error: Error) => {
    if (doneDispatched) return
    doneDispatched = true
    options.onError(error)
  }

  fetch(`${getAgentStreamBaseURL()}/v1/${path}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        // Try to parse error body
        let msg = `HTTP ${response.status}`
        try {
          const errBody = await response.json()
          msg = errBody?.msg || errBody?.message || msg
        } catch {
          // ignore parse errors
        }
        dispatchStreamError(new Error(msg))
        return
      }

      if (!response.body) {
        dispatchStreamError(new Error(options.noBodyMessage))
        return
      }

      const responseSessionId = response.headers.get('X-Agent-Session-ID')
      if (responseSessionId) {
        options.onSessionId?.(responseSessionId)
      }

      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      const processBuffer = () => {
        // SSE blocks are separated by double newlines
        const blocks = buffer.split('\n\n')
        // Keep the last (potentially incomplete) block in the buffer
        buffer = blocks.pop() ?? ''

        for (const block of blocks) {
          if (!block.trim()) continue

          let eventType = 'message'
          let dataStr = ''

          for (const line of block.split('\n')) {
            if (line.startsWith('event:')) {
              eventType = line.slice(6).trim()
            } else if (line.startsWith('data:')) {
              dataStr = line.slice(5).trim()
            }
          }

          if (!dataStr) continue

          let parsed: unknown
          try {
            parsed = JSON.parse(dataStr)
          } catch {
            parsed = dataStr
          }

          const sseEvent: AgentSSEEvent = {
            event: eventType as AgentSSEEvent['event'],
            data: parsed,
          }

          if (sseEvent.event === 'done') {
            dispatchDone()
            return
          }

          if (sseEvent.event === 'error') {
            dispatchStreamError(new Error(getAgentStreamError(parsed, options.fallbackStreamError)))
            return
          }

          options.onEvent(sseEvent)
        }
      }

      const pump = async (): Promise<void> => {
        try {
          const { done, value } = await reader.read()
          if (done) {
            // Flush remaining buffer
            if (buffer.trim()) {
              processBuffer()
            }
            dispatchDone()
            return
          }
          buffer += decoder.decode(value, { stream: true })
          processBuffer()
          return pump()
        } catch (err) {
          if ((err as Error)?.name === 'AbortError') {
            // Silently ignore user-initiated aborts
            return
          }
          dispatchStreamError(err instanceof Error ? err : new Error(String(err)))
        }
      }

      await pump()
    })
    .catch((err) => {
      if (err?.name === 'AbortError') return
      dispatchStreamError(err instanceof Error ? err : new Error(String(err)))
    })

  return controller
}

/**
 * Connect to the agent chat endpoint via SSE (fetch + ReadableStream).
 * Returns an AbortController that can be used to cancel the stream.
 */
export function connectAgentChat(
  sessionId: string | null,
  requestId: string,
  message: string,
  pageContext: AgentPageContext,
  orchestrationMode: 'single_agent',
  clientContext: { locale?: string; timezone?: string } | undefined,
  onEvent: (event: AgentSSEEvent) => void,
  onError: (error: Error) => void,
  onDone: () => void,
  onSessionId?: (sessionId: string) => void
): AbortController {
  return connectAgentStream<AgentChatRequest>(
    'agent/chat',
    {
      sessionId,
      requestId,
      message,
      pageContext,
      orchestrationMode,
      clientContext,
    },
    {
      noBodyMessage: 'No response body for SSE stream',
      fallbackStreamError: 'Agent error',
      onEvent,
      onError,
      onDone,
      onSessionId,
    }
  )
}

export function connectAgentAskStream(
  sessionId: string | null,
  requestId: string,
  message: string,
  pageContext: AgentChatRequest['pageContext'],
  clientContext: AgentChatRequest['clientContext'] | undefined,
  jobName: string | undefined,
  onEvent: (event: AgentSSEEvent) => void,
  onError: (error: Error) => void,
  onDone: () => void,
  onSessionId?: (sessionId: string) => void
): AbortController {
  return connectAgentStream<AgentAskRequest>(
    'agent/ask/stream',
    {
      sessionId,
      requestId,
      message,
      jobName,
      pageContext,
      clientContext,
    },
    {
      noBodyMessage: 'No response body for ask stream',
      fallbackStreamError: 'ask error',
      onEvent,
      onError,
      onDone,
      onSessionId,
    }
  )
}

export function connectAgentResume(
  confirmId: string,
  onEvent: (event: AgentSSEEvent) => void,
  onError: (error: Error) => void,
  onDone: () => void,
  onSessionId?: (sessionId: string) => void
): AbortController {
  return connectAgentStream<AgentResumeRequest>(
    'agent/chat/resume',
    { confirmId },
    {
      noBodyMessage: 'No response body for SSE stream',
      fallbackStreamError: 'Agent error',
      onEvent,
      onError,
      onDone,
      onSessionId,
    }
  )
}

// ──────────────────────────────────────────────
// REST endpoints
// ──────────────────────────────────────────────

/**
 * Confirm or reject a pending write-operation.
 */
export const apiConfirmAction = (
  confirmId: string,
  confirmed: boolean,
  payload?: Record<string, unknown>
) =>
  apiV1Post<IResponse<AgentConfirmResponseData>>('agent/chat/confirm', {
    confirmId,
    confirmed,
    payload,
  })

/**
 * List agent sessions for the current user and current surface.
 */
export const apiListSessions = (surface?: AgentSurface) =>
  apiV1Get<IResponse<AgentSession[]>>('agent/sessions', {
    searchParams: surface ? { surface } : undefined,
  })

export const apiPinSession = (sessionId: string, pinned: boolean) =>
  apiV1Put<IResponse<AgentSession>>(`agent/sessions/${encodeURIComponent(sessionId)}/pin`, {
    pinned,
  })

export const apiRenameSession = (sessionId: string, title: string) =>
  apiV1Put<IResponse<AgentSession>>(`agent/sessions/${encodeURIComponent(sessionId)}/title`, {
    title,
  })

export const apiDeleteSession = (sessionId: string) =>
  apiV1Delete<IResponse<string>>(`agent/sessions/${encodeURIComponent(sessionId)}`)

/**
 * Fetch the message history for a given session.
 */
export const apiGetSessionMessages = (sessionId: string) =>
  apiV1Get<IResponse<AgentMessage[]>>(`agent/sessions/${encodeURIComponent(sessionId)}/messages`)

export const apiGetSessionToolCalls = (sessionId: string) =>
  apiV1Get<IResponse<AgentToolCall[]>>(`agent/sessions/${encodeURIComponent(sessionId)}/tool-calls`)

export const apiGetSessionTurns = (sessionId: string) =>
  apiV1Get<IResponse<AgentTurn[]>>(`agent/sessions/${encodeURIComponent(sessionId)}/turns`)

export const apiGetTurnEvents = (turnId: string) =>
  apiV1Get<IResponse<AgentEvent[]>>(`agent/turns/${encodeURIComponent(turnId)}/events`)

export const apiGetAgentConfigSummary = () =>
  apiV1Get<IResponse<AgentConfigSummary>>('agent/config-summary')

/**
 * Submit parameter review result (confirm or modify).
 */
export const apiParameterUpdate = (
  sessionId: string,
  reviewId: string,
  action: 'confirm' | 'modify',
  parameters: Record<string, unknown>
) =>
  apiV1Post<IResponse<{ status: string }>>('agent/chat/parameter-update', {
    session_id: sessionId,
    review_id: reviewId,
    action,
    parameters,
  })

// ──────────────────────────────────────────────
// Feedback
// ──────────────────────────────────────────────

export interface AgentFeedback {
  id: number
  sessionId: string
  userId: number
  accountId: number
  targetType: 'message' | 'turn'
  targetId: string
  rating: 1 | -1
  tags?: string[]
  dimensions?: Record<string, number>
  comment?: string
  status: 'draft' | 'submitted'
  submittedAt?: string | null
  createdAt: string
  updatedAt: string
}

export interface FeedbackUpsertRequest {
  sessionId: string
  targetType: 'message' | 'turn'
  targetId: string
  rating: 1 | -1
  tags?: string[]
  dimensions?: Record<string, number>
  comment?: string
}

export const apiUpsertFeedback = (req: FeedbackUpsertRequest) =>
  apiV1Put<IResponse<AgentFeedback>>('agent/feedbacks', req)

export const apiSubmitFeedback = (sessionId: string, targetType: string, targetId: string) =>
  apiV1Post<IResponse<AgentFeedback>>('agent/feedbacks/submit', {
    sessionId,
    targetType,
    targetId,
  })

export const apiQuickSubmitFeedback = (req: {
  sessionId: string
  targetType: 'message' | 'turn'
  targetId: string
  rating: 1 | -1
  tags?: string[]
  dimensions?: Record<string, number>
  comment?: string
}) => apiV1Post<IResponse<AgentFeedback>>('agent/feedbacks/quick-submit', req)

export const apiEnrichFeedback = (req: {
  sessionId: string
  targetType: 'message' | 'turn'
  targetId: string
  tags?: string[]
  dimensions?: Record<string, number>
  comment?: string
}) => apiV1Put<IResponse<AgentFeedback>>('agent/feedbacks/enrich', req)

export const apiListFeedbacks = (sessionId: string) =>
  apiV1Get<IResponse<AgentFeedback[]>>(`agent/feedbacks?sessionId=${encodeURIComponent(sessionId)}`)

File: backend/dao/model/agent.go
package model

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// AgentSession represents a conversation session between a user and the agent.
type AgentSession struct {
	ID                    uint           `gorm:"primarykey" json:"id"`
	SessionID             string         `gorm:"type:uuid;uniqueIndex;not null" json:"sessionId"`
	UserID                uint           `gorm:"not null" json:"userId"`
	AccountID             uint           `gorm:"not null" json:"accountId"`
	Title                 string         `gorm:"type:varchar(255)" json:"title"`
	Source                string         `gorm:"type:varchar(32);not null;default:'chat';index" json:"source"` // chat | ops_audit | system | benchmark
	PageContext           datatypes.JSON `json:"pageContext"`
	MessageCount          int            `gorm:"default:0" json:"messageCount"`
	LastOrchestrationMode string         `gorm:"type:varchar(32);default:'single_agent'" json:"lastOrchestrationMode"`
	PinnedAt              *time.Time     `gorm:"index" json:"pinnedAt,omitempty"`
	CreatedAt             time.Time      `json:"createdAt"`
	UpdatedAt             time.Time      `json:"updatedAt"`
	DeletedAt             gorm.DeletedAt `gorm:"index" json:"deletedAt"`
}

// AgentMessage represents an individual message in an agent session.
type AgentMessage struct {
	ID         uint           `gorm:"primarykey" json:"id"`
	SessionID  string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	Role       string         `gorm:"type:varchar(20);not null" json:"role"` // user|assistant|tool
	Content    string         `gorm:"type:text" json:"content"`
	ToolCalls  datatypes.JSON `json:"toolCalls,omitempty"`
	ToolCallID string         `gorm:"type:varchar(100)" json:"toolCallId,omitempty"`
	ToolName   string         `gorm:"type:varchar(100)" json:"toolName,omitempty"`
	Metadata   datatypes.JSON `json:"metadata,omitempty"`
	CreatedAt  time.Time      `json:"createdAt"`
}

// AgentToolCall is an audit log for tool executions, also used for benchmark data collection.
type AgentToolCall struct {
	ID               uint           `gorm:"primarykey" json:"id"`
	SessionID        string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	TurnID           string         `gorm:"type:uuid;index" json:"turnId,omitempty"`
	MessageID        *uint          `json:"messageId,omitempty"`
	ToolCallID       string         `gorm:"type:varchar(128);index" json:"toolCallId,omitempty"`
	AgentID          string         `gorm:"type:varchar(128);index" json:"agentId,omitempty"`
	ParentEventID    *uint          `gorm:"index" json:"parentEventId,omitempty"`
	AgentRole        string         `gorm:"type:varchar(32);index" json:"agentRole,omitempty"`
	Source           string         `gorm:"type:varchar(32);not null;default:'backend';index" json:"source"` // backend | local | benchmark
	ToolName         string         `gorm:"type:varchar(100);not null;index" json:"toolName"`
	ToolArgs         datatypes.JSON `gorm:"not null" json:"toolArgs"`
	ToolResult       datatypes.JSON `json:"toolResult,omitempty"`
	ResultStatus     string         `gorm:"type:varchar(32);not null;default:'success'" json:"resultStatus"`
	ExecutionBackend string         `gorm:"type:varchar(64)" json:"executionBackend,omitempty"`
	LatencyMs        int            `json:"latencyMs,omitempty"`
	TokenCount       int            `json:"tokenCount,omitempty"`
	UserConfirmed    *bool          `json:"userConfirmed,omitempty"`
	CreatedAt        time.Time      `json:"createdAt"`
}

// AgentTurn represents one agent execution turn within a session.
type AgentTurn struct {
	ID                uint           `gorm:"primarykey" json:"id"`
	TurnID            string         `gorm:"type:uuid;uniqueIndex;not null" json:"turnId"`
	SessionID         string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	RequestID         string         `gorm:"type:varchar(128);index" json:"requestId,omitempty"`
	OrchestrationMode string         `gorm:"type:varchar(32);default:'single_agent';index" json:"orchestrationMode"`
	RootAgentID       string         `gorm:"type:varchar(128);index" json:"rootAgentId,omitempty"`
	Status            string         `gorm:"type:varchar(32);default:'running';index" json:"status"`
	FinalMessageID    *uint          `gorm:"index" json:"finalMessageId,omitempty"`
	Metadata          datatypes.JSON `json:"metadata,omitempty"`
	StartedAt         time.Time      `gorm:"index" json:"startedAt"`
	EndedAt           *time.Time     `json:"endedAt,omitempty"`
	CreatedAt         time.Time      `json:"createdAt"`
	UpdatedAt         time.Time      `json:"updatedAt"`
}

// AgentRunEvent stores semantic events emitted during an agent turn.
type AgentRunEvent struct {
	ID            uint           `gorm:"primarykey" json:"id"`
	TurnID        string         `gorm:"type:uuid;index;not null" json:"turnId"`
	SessionID     string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	AgentID       string         `gorm:"type:varchar(128);index" json:"agentId"`
	ParentAgentID string         `gorm:"type:varchar(128);index" json:"parentAgentId,omitempty"`
	AgentRole     string         `gorm:"type:varchar(32);index" json:"agentRole"`
	EventType     string         `gorm:"type:varchar(64);index" json:"eventType"`
	EventStatus   string         `gorm:"type:varchar(32);index" json:"eventStatus,omitempty"`
	Title         string         `gorm:"type:varchar(255)" json:"title,omitempty"`
	Content       string         `gorm:"type:text" json:"content,omitempty"`
	Metadata      datatypes.JSON `json:"metadata,omitempty"`
	Sequence      int            `gorm:"index" json:"sequence"`
	StartedAt     *time.Time     `json:"startedAt,omitempty"`
	EndedAt       *time.Time     `json:"endedAt,omitempty"`
	CreatedAt     time.Time      `json:"createdAt"`
}

// AgentFeedback stores user feedback (thumbs up/down + optional details) for a message or turn.
type AgentFeedback struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	SessionID   string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	UserID      uint           `gorm:"not null;uniqueIndex:idx_feedback_unique,priority:1" json:"userId"`
	AccountID   uint           `gorm:"not null;index" json:"accountId"`
	TargetType  string         `gorm:"type:varchar(16);not null;uniqueIndex:idx_feedback_unique,priority:2" json:"targetType"` // message | turn
	TargetID    string         `gorm:"type:varchar(128);not null;uniqueIndex:idx_feedback_unique,priority:3" json:"targetId"`  // message.id or turn_id
	Rating      int16          `gorm:"not null" json:"rating"`                                                                 // 1 = thumbs up, -1 = thumbs down
	Tags        datatypes.JSON `json:"tags,omitempty"`                                                                         // ["inaccurate","helpful",...]
	Dimensions  datatypes.JSON `json:"dimensions,omitempty"`                                                                   // {"relevance":4,"accuracy":3,...}
	Comment     string         `gorm:"type:text" json:"comment,omitempty"`
	Status      string         `gorm:"type:varchar(16);not null;default:'draft'" json:"status"` // draft | submitted
	SubmittedAt *time.Time     `json:"submittedAt,omitempty"`
	EnrichedAt  *time.Time     `json:"enrichedAt,omitempty"`
	CreatedAt   time.Time      `json:"createdAt"`
	UpdatedAt   time.Time      `json:"updatedAt"`
}

// AgentQualityEval stores LLM-as-Judge quality evaluation results for agent chat sessions.
// eval_scope: 'session' | 'turn'
// eval_type: 'full' | 'dialogue' | 'task'
// trigger_source: 'feedback' (user-triggered) | 'offline_batch' | 'manual'
// eval_status: 'pending' | 'running' | 'completed' | 'failed'
type AgentQualityEval struct {
	ID            uint           `gorm:"primarykey" json:"id"`
	SessionID     string         `gorm:"type:uuid;index;not null" json:"sessionId"`
	TurnID        string         `gorm:"type:uuid;index" json:"turnId,omitempty"`
	EvalScope     string         `gorm:"type:varchar(16);not null;default:'session';index" json:"evalScope"`
	EvalType      string         `gorm:"type:varchar(32);not null;default:'full';index" json:"evalType"`
	TargetID      string         `gorm:"type:varchar(128);index" json:"targetId,omitempty"`
	FeedbackID    *uint          `gorm:"index" json:"feedbackId,omitempty"`
	TriggerSource string         `gorm:"type:varchar(32);not null;index" json:"triggerSource"`
	EvalStatus    string         `gorm:"type:varchar(16);not null;default:'pending';index" json:"evalStatus"`
	ChatScores    datatypes.JSON `json:"chatScores,omitempty"`
	ChainScores   datatypes.JSON `json:"chainScores,omitempty"`
	ChatModel     string         `gorm:"type:varchar(64)" json:"chatModel,omitempty"`
	ChainModel    string         `gorm:"type:varchar(64)" json:"chainModel,omitempty"`
	Summary       string         `gorm:"type:text" json:"summary,omitempty"`
	RawChatResp   datatypes.JSON `json:"rawChatResp,omitempty"`
	RawChainResp  datatypes.JSON `json:"rawChainResp,omitempty"`
	ArtifactPath  string         `gorm:"type:text" json:"artifactPath,omitempty"`
	Metadata      datatypes.JSON `json:"metadata,omitempty"`
	CreatedAt     time.Time      `gorm:"not null;default:now()" json:"createdAt"`
	CompletedAt   *time.Time     `json:"completedAt,omitempty"`
	UpdatedAt     time.Time      `gorm:"not null;default:now()" json:"updatedAt"`
}

// JobLogSnapshot is a persisted log snippet captured when a job reaches terminal state.
type JobLogSnapshot struct {
	ID            uint      `gorm:"primarykey" json:"id"`
	JobName       string    `gorm:"type:varchar(255);index;not null" json:"jobName"`
	PodName       string    `gorm:"type:varchar(255);not null" json:"podName"`
	ContainerName string    `gorm:"type:varchar(255);not null" json:"containerName"`
	LogTail       string    `gorm:"type:text" json:"logTail"`
	LogHead       string    `gorm:"type:text" json:"logHead"`
	CapturedAt    time.Time `gorm:"not null" json:"capturedAt"`
	JobStatus     string    `gorm:"type:varchar(50)" json:"jobStatus"`
	CreatedAt     time.Time `json:"createdAt"`
}
